### PR TITLE
Added needsreauthentication to core user authentication endpoint.

### PR DESCRIPTION
**Pull request by ShahAaron**
## Summary

Addresses issue #1559 

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/googlesitekit/modules/create-info-store.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+		 * Returns admin screen url.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param {Object} state Data store's state.
+		 * @param {(Object|undefined)} queryArgs Query arguments to add to admin URL.
+		 * @return {(string|undefined)} The admin screen url.
+		 */
+		getAdminScreenURL: createRegistrySelector( ( select ) => ( state, queryArgs ) => {
+			return select( CORE_SITE ).getAdminURL( adminPage, queryArgs );
```
Comment (tofumatt) [inline assets/js/googlesitekit/modules/create-info-store.js:65]: Passing `queryArgs` here won't work as-expected, because the first argument for `getAdminURL` is `page`. So this is going to pass the query args to the wrong argument. Since `page` is an optional argument, it looks like you want:

```suggestion
			return select( CORE_SITE ).getAdminURL( adminPage, undefined, queryArgs );
```

File: assets/js/googlesitekit/modules/create-info-store.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+		/**
+		 * Returns admin reauth url.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param {Object} state Data store's state.
+		 * @param {(Object|undefined)} queryArgs Query arguments to add to admin URL.
+		 * @return {(string|undefined)} The admin reauth url.
+		 */
+		getAdminReauthURL: createRegistrySelector( ( select ) => ( state, queryArgs ) => {
```
Comment (tofumatt) [inline assets/js/googlesitekit/modules/create-info-store.js:77]: There aren't any mention of `queryArgs` for this selector in the IB, so I think you can remove it, from my understanding.
```suggestion
		getAdminReauthURL: createRegistrySelector( ( select ) => ( state ) => {
```

File: assets/js/googlesitekit/modules/create-info-store.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+			const noSetupQueryArgs = ! requiresSetup ? {
+				notification: 'authentication_success',
+				reAuth: undefined,
+			} : {};
+
+			if ( ! needsReauthentication ) {
+				return select( CORE_SITE ).getAdminURL( adminPage, noSetupQueryArgs );
+			}
+
+			return select( CORE_SITE ).getAdminURL( adminPage, queryArgs );
```
Comment (tofumatt) [inline assets/js/googlesitekit/modules/create-info-store.js:89]: If removed above, you'll need to remove the args here too (fwiw, these also would have went to the `page` arg, not the extra `queryArgs` argument):

```suggestion
			return select( CORE_SITE ).getAdminURL( adminPage );
```

File: assets/js/googlesitekit/modules/create-info-store.js (hunk 1 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/modules/create-info-store.js]: Done

File: assets/js/googlesitekit/modules/create-info-store.js (hunk 2 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/modules/create-info-store.js]: Done

File: assets/js/googlesitekit/modules/create-info-store.js (hunk 3 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/modules/create-info-store.js]: Fixed. I've changed this to use getAdminScreenURL

File: assets/js/googlesitekit/modules/create-info-store.js (hunk 1 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/modules/create-info-store.js]: @tofumatt could I confirm this with you. The first argument is in fact `page` but I'm passing `adminPage` to that, so the second argument should be `queryArgs` not undefined otherwise no args get passed.

File: assets/js/googlesitekit/datastore/user/authentication.js [hunk 4]
```diff
@@ -239,6 +239,22 @@ export const selectors = {
...
 		return unsatisfiedScopes;
 	} ),
+
+	/**
+	 * Checks reauthentication status for this user.
+	 *
+	 * Returns true if any required scopes are not satisfied or undefined
+	 * if reauthentication info is not available/loaded.
+	 *
+	 * @since 1.9.0
```
Comment (felixarntz) [inline assets/js/googlesitekit/datastore/user/authentication.js:249]: Should be `@since n.e.x.t`, exact version numbers will always be added as part of the release preparation.

File: assets/js/googlesitekit/datastore/user/authentication.test.js [hunk 5]
```diff
@@ -430,5 +431,50 @@ describe( 'core/user authentication', () => {
 				expect( unsatisfiedScopes ).toEqual( undefined );
 			} );
 		} );
+
+		describe( 'needsReauthentication', () => {
```
Comment (felixarntz) [inline assets/js/googlesitekit/datastore/user/authentication.test.js:435]: This selector should also have the "uses a resolver get all authentication info" test like the other ones do.

File: assets/js/googlesitekit/modules/create-info-store.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+const { createRegistrySelector } = Data;
+
+/**
+ * Creates a store object that has selectors for managing site info.
+ *
+ * @since 1.6.0
+ * @private
+ *
+ * @param {Object}  options               Options to consider for the store.
+ * @param {number}  options.storeName     Store name to use.
```
Comment (felixarntz) [inline assets/js/googlesitekit/modules/create-info-store.js:35]: Since the store name here is required for the logic below to work, I'd argue it should have a default based on a required `slug` parameter (like `createModuleStore`). We'll also need this `slug` parameter further below in `getAdminReauthURL` anyway (see other comment).

File: assets/js/googlesitekit/modules/create-info-store.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+ * @param {string}  options.adminPage     Store admin page. Default is 'googlesitekit-dashboard'.
+ * @param {boolean} options.requiresSetup Store flag, for requires setup. Default is 'true'
+ * @return {Object} The info store object.
+ */
+export const createInfoStore = ( {
+	storeName = undefined,
+	adminPage = undefined,
+	requiresSetup = undefined,
```
Comment (felixarntz) [inline assets/js/googlesitekit/modules/create-info-store.js:43]: These two arguments are missing their defaults here.

File: assets/js/googlesitekit/modules/create-info-store.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+	};
+	const resolvers = {};
+	const selectors = {
+		/**
+		 * Returns admin screen url.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param {(Object|undefined)} queryArgs Query arguments to add to admin URL.
+		 * @return {(string|undefined)} The admin screen url.
```
Comment (felixarntz) [inline assets/js/googlesitekit/modules/create-info-store.js:61]:
```suggestion
		 * @return {(string|undefined)} The admin screen URL.
```

File: assets/js/googlesitekit/modules/create-info-store.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+	const INITIAL_STATE = {};
+	const actions = {};
+	const controls = {};
+	const reducer = ( state ) => {
+		return { ...state };
+	};
+	const resolvers = {};
+	const selectors = {
+		/**
+		 * Returns admin screen url.
```
Comment (felixarntz) [inline assets/js/googlesitekit/modules/create-info-store.js:56]:
```suggestion
		 * Returns admin screen URL.
```

File: assets/js/googlesitekit/modules/create-info-store.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+		 *
+		 * @param {(Object|undefined)} queryArgs Query arguments to add to admin URL.
+		 * @return {(string|undefined)} The admin screen url.
+		 */
+		getAdminScreenURL: createRegistrySelector( ( select ) => ( queryArgs ) => {
+			return select( CORE_SITE ).getAdminURL( adminPage, queryArgs );
+		} ),
+
+		/**
+		 * Returns admin reauth url.
```
Comment (felixarntz) [inline assets/js/googlesitekit/modules/create-info-store.js:68]:
```suggestion
		 * Returns admin reauthentication URL.
```

File: assets/js/googlesitekit/modules/create-info-store.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @return {(string|undefined)} The admin reauth url.
```
Comment (felixarntz) [inline assets/js/googlesitekit/modules/create-info-store.js:72]:
```suggestion
		 * @return {(string|undefined)} The admin reauthentication URL.
```

File: assets/js/googlesitekit/modules/create-info-store.js [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+			const noSetupQueryArgs = ! requiresSetup ? {
+				notification: 'authentication_success',
+				reAuth: undefined,
+			} : {};
+
+			if ( ! needsReauthentication ) {
+				return select( STORE_NAME ).getAdminScreenURL( noSetupQueryArgs );
+			}
+
+			return select( CORE_SITE ).getAdminScreenURL();
```
Comment (felixarntz) [inline assets/js/googlesitekit/modules/create-info-store.js:86]: The logic in this method is not quite right. It needs to match _exactly_ what the current `getReAuthURL()` utility function does, except having the specifics being applied based on `adminPage` and `requiresSetup`. Some issues here are:
* Missing query parameters `slug` and `reAuth` (the `status` value from the original `getReAuthURL` function should always be `true` here, since we only use this URL for module activations as of now, never deactivations).
* Missing usage of the `connectURL`: The URL that's being built here should be like the `redirect` variable in the original `getReAuthURL` function; if `! needsReauthentication`, it should be returned as is, otherwise it should be added as `redirect` query parameter to the `connectURL` (with similar encoding etc. like in the original function), alongside `status: true`.
* This function should never return the general `select( CORE_SITE ).getAdminScreenURL()`.

File: assets/js/googlesitekit/modules/create-module-store.js [hunk 13]
```diff
@@ -68,16 +73,25 @@ export const createModuleStore = ( slug, {
 		storeName,
 	} );
 
+	const infoStore = createInfoStore( {
```
Comment (felixarntz) [inline assets/js/googlesitekit/modules/create-module-store.js:76]: See other comment, let's pass `slug` here as first parameter.

File: assets/js/modules/adsense/settings/settings-setup-incomplete.js [hunk 14]
```diff
@@ -48,7 +47,7 @@ export default function SettingsSetupIncomplete() {
 				className="googlesitekit-settings-module__edit-button"
 				href={
 					// TODO: Replace this function with something datastore-driven.
-					getReAuthURL( 'adsense', true )
+					useSelect( ( select ) => select( STORE_NAME ).getAdminReauthURL() )
```
Comment (felixarntz) [inline assets/js/modules/adsense/settings/settings-setup-incomplete.js:50]: This should be better assigned to a variable further above with the other hooks instead of inline here.

File: assets/js/modules/pagespeed-insights/settings/settings-view.js [hunk 15]
```diff
@@ -26,11 +26,11 @@ import { __, sprintf } from '@wordpress/i18n';
  */
 import Data from 'googlesitekit-data';
 import { sanitizeHTML } from '../../../util';
-import { STORE_NAME as CORE_SITE } from '../../../googlesitekit/datastore/site/constants';
+import { STORE_NAME } from '../datastore/constants';
 const { useSelect } = Data;
 
 export default function SettingsView() {
-	const dashboardPermalink = useSelect( ( select ) => select( CORE_SITE ).getAdminURL( 'googlesitekit-dashboard' ) );
+	const dashboardPermalink = useSelect( ( select ) => select( STORE_NAME ).getAdminReauthURL() );
```
Comment (felixarntz) [inline assets/js/modules/pagespeed-insights/settings/settings-view.js:33]: This should not use `getAdminReauthURL`, but `getAdminScreenURL` - it's just for viewing the data, not for setting up the module.

File: assets/js/modules/search-console/datastore/index.js [hunk 16]
```diff
@@ -29,6 +29,7 @@ const baseModuleStore = Modules.createModuleStore( 'search-console', {
 	settingSlugs: [
 		'propertyID',
 	],
+	adminPage: 'googlesitekit-module-search-console',
```
Comment (felixarntz) [inline assets/js/modules/search-console/datastore/index.js:32]: This should also receive `requiresSetup: false`, since Search Console does not have any setup in the plugin.

File: package-lock.json [hunk 17]
```diff
@@ -35482,7 +35482,7 @@
       "dependencies": {
         "strip-ansi": {
           "version": "3.0.1",
-          "resolved": "http://registry.npmjs.org/strip-ansi/-/strip-ansi-3.0.1.tgz",
+          "resolved": "https://registry.npmjs.org/strip-ansi/-/strip-ansi-3.0.1.tgz",
```
Comment (felixarntz) [inline package-lock.json:35485]: The changes in this file seem irrelevant, so that should be reverted?

**Comment by tofumatt** [comment]
@googlebot I consent.

File: assets/js/googlesitekit/modules/create-info-store.js (hunk 6 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/modules/create-info-store.js:36]: Slug added and defaults for `storename`

File: assets/js/modules/search-console/datastore/index.js (hunk 16 above)
Comment (ShahAaron) [reply assets/js/modules/search-console/datastore/index.js:32]: Added with default to false

File: assets/js/googlesitekit/datastore/user/authentication.js (hunk 4 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/datastore/user/authentication.js]: Done

File: assets/js/googlesitekit/modules/create-info-store.js (hunk 7 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/modules/create-info-store.js]: Defaults added

File: assets/js/googlesitekit/modules/create-info-store.js (hunk 8 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/modules/create-info-store.js]: Done

File: assets/js/googlesitekit/modules/create-info-store.js (hunk 9 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/modules/create-info-store.js]: Done

File: assets/js/googlesitekit/modules/create-info-store.js (hunk 10 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/modules/create-info-store.js]: Done

File: assets/js/googlesitekit/modules/create-info-store.js (hunk 11 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/modules/create-info-store.js]: Done

File: assets/js/googlesitekit/modules/create-module-store.js (hunk 13 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/modules/create-module-store.js]: Slug added to parameters

File: assets/js/modules/adsense/settings/settings-setup-incomplete.js (hunk 14 above)
Comment (ShahAaron) [reply assets/js/modules/adsense/settings/settings-setup-incomplete.js]: Variable added - adminReauthURL

File: assets/js/modules/pagespeed-insights/settings/settings-view.js (hunk 15 above)
Comment (ShahAaron) [reply assets/js/modules/pagespeed-insights/settings/settings-view.js]: Done

File: assets/js/googlesitekit/datastore/user/authentication.test.js (hunk 5 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/datastore/user/authentication.test.js:435]: Test added

File: assets/js/googlesitekit/modules/create-info-store.js (hunk 12 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/modules/create-info-store.js]: @felixarntz could I clarify this change with you. 

I've added `reAuth` to the parameter list for `getAdminReauthURL`, status is obtained as it is passed into the store. These 2 arguments are then passed to `getAdminScreenURL`.

My question surrounds the use of `connectURL or `adminURL` as the original function `getReAuthURL` required these as it was building the entire URL string but the new `getAdminURL` doesn't require them as it builds the base url and and adds the args to it. So my question is do we actually need `connectURL or `adminURL`?

Comment (felixarntz) [reply assets/js/googlesitekit/modules/create-info-store.js]: @ShahAaron We need `getAdminURL` to build the URL to the actual module page (or dashboard). But then, if `needsReauthentication` is `true`, we also need the connect URL (see `getReAuthURL` function's last few lines). There's the `getConnectURL` selector in `core/user` to retrieve that, and we'd need to pass the original URL built as `redirectURL` argument there.

**Comment by felixarntz** [comment]
@ShahAaron Something minor I just thought about that we'll need to do in this PR as well is add `googlesitekit-datastore-site` and `googlesitekit-datastore-user` as dependencies to the `googlesitekit-modules` asset in PHP, in the `Assets` class. That's necessary since now `createModuleStore` which is part of `googlesitekit-modules` relies on data from `core/site` and `core/user`.

File: assets/js/googlesitekit/modules/create-info-store.js (hunk 12 above)
Comment (ShahAaron) [reply assets/js/googlesitekit/modules/create-info-store.js]: I have refactored this to include `connectURL` where the `needsReauthentication` is true
